### PR TITLE
Improve performance of hover tool and tooltips

### LINKED ISSUE #10234: Improve responsiveness when hovering over many points
#### My Pain

I would like to use Bokeh plots in my applications but sometimes they feel slow and sluggish.

An example of this is plots with many points. The many points by them selves is not the main problem. The main problem is hovering.

If the users hovers over the Bokeh plot it becomes unresponsive and my fan turns on. 

I suspect its due to collecting and showing tool tips for many points

My fear is that my users turn away from my apps because they appear slow or dead.

#### Example

https://ferret.pmel.noaa.gov/osmc/dashboard

I hover and the drag the plot. The first time its OK responsive. The second time it takes 10 secs before the plot moves.

bokeh-slow

#### Another Example

Its the same problem at awesome-panel.org in the Kickstarter Dashboard. If I want to select some points to filter on the hovering often steps in first and I have to wait 5-10 secs for the plot to respond.

image

#### Solution

Fix this :-) For example by 

- fixing the cause if there is a cause or
- setting a max on the number of points to show hover information for.

FYI @philippjfr

### PR DESCRIPTION
**Pull request by mattpap**
This improves various aspects of generating HTML tooltip content by hover tool. 

fixes #10234

### CODE REVIEW

**Comment by mattpap** [comment]
This reduces CPU utilization significantly (several fold). Further optimizations are possible, but I will leave this for future work.

File: bokehjs/src/lib/core/util/templating.ts [hunk 1]
```diff
@@ -113,11 +113,11 @@ export function get_value(raw_name: string, data_source: ColumnarDataSource, i:
   }
 }
 
-export function replace_placeholders(str: string, data_source: ColumnarDataSource, i: Index, formatters?: Formatters, special_vars: Vars = {}): string {
-
+export function replace_placeholders(str: string, data_source: ColumnarDataSource, i: Index, formatters?: Formatters, special_vars: Vars = {}): string | Node[]  {
   // this handles the special case @$name, replacing it with an @var corresponding to special_vars.name
   str = str.replace(/@\$name/g, (_match) => `@{${special_vars.name}}`)
 
+  let unsafe = false
```
Comment (bryevdv) [inline bokehjs/src/lib/core/util/templating.ts:120]: For a user, "safe" as a format means "I am *stipulating* this content is safe (to me) and therefore does not require escaping". Using the name "unsafe" here to also mean "no escaping" seems potentially confusing. How about 
```
let unescaped = false  // or escaped = true
```

 instead?

File: bokeh/models/annotations.py [hunk 2]
```diff
@@ -1048,6 +1048,10 @@ class Tooltip(Annotation):
 
     inner_only = Bool(default=True, help="""
     Whether to display outside a central plot frame area.
+
+    .. note:
+        This property is deprecated and will be removed in bokeh 3.0.
```
Comment (bryevdv) [inline bokeh/models/annotations.py:1053]: Can you explain this more? Normally we don't want hover tooltips to show unless the cursor is inside the plot frame. Are you just making that the unconditional behavior all the time?

Comment (bryevdv) [reply bokeh/models/annotations.py:1053]: Also just to reiterate: I would like to deprecate the entire `ToolTip` model. It has a bad name, and importantly , it's never actually been usable by users, so this should not be a stretch.

Comment (mattpap) [reply bokeh/models/annotations.py:1053]: Now `HoverTool` decides whether a tooltip should be visible or not, decoupling it from the frame (for the most part).

File: bokehjs/src/lib/core/util/templating.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/src/lib/core/util/templating.ts]: I changed it to `has_html`, as it is what this variable really conveys.

**Pull request merged**